core/src/main/scala/kafka/server/metadata/ZkMetadataCache.scala

```diff
@@ -430,4 +451,103 @@ class ZkMetadataCache(brokerId: Int) extends MetadataCache with Logging {
                               aliveNodes: mutable.LongMap[collection.Map[ListenerName, Node]]) {
     val topicNames: Map[Uuid, String] = topicIds.map { case (topicName, topicId) => (topicId, topicName) }
   }
+
+  override def metadataVersion(): MetadataVersion = metadataVersion
+
+  override def features(): FinalizedFeaturesAndEpoch = {
+    featuresAndEpoch match {
+      case Some(features) => features
+      case None => FinalizedFeaturesAndEpoch(Map.empty, ApiVersionsResponse.UNKNOWN_FINALIZED_FEATURES_EPOCH)
+    }
+  }
+
+  /**
+   * Updates the cache to the latestFeatures, and updates the existing epoch to latestEpoch.
+   * Expects that the latestEpoch should be always greater than the existing epoch (when the
+   * existing epoch is defined).
+   *
+   * @param latestFeatures   the latest finalized features to be set in the cache
+   * @param latestEpoch      the latest epoch value to be set in the cache
+   *
+   * @throws                 FeatureCacheUpdateException if the cache update operation fails
+   *                         due to invalid parameters or incompatibilities with the broker's
+   *                         supported features. In such a case, the existing cache contents are
+   *                         not modified.
+   */
+  def updateFeaturesOrThrow(latestFeatures: Map[String, Short], latestEpoch: Long): Unit = {
+    val latest = FinalizedFeaturesAndEpoch(latestFeatures, latestEpoch)
+    val existing = featuresAndEpoch.map(item => item.toString()).getOrElse("<empty>")
+    if (featuresAndEpoch.isDefined && featuresAndEpoch.get.epoch > latest.epoch) {
+      val errorMsg = s"FinalizedFeatureCache update failed due to invalid epoch in new $latest." +
```

Comment:
Do we really need to have this here? The epoch thing just seems like another example of overdesign

It's basically validating that ZK is working as expected...

I could see this getting annoying if we wanted to manually go in and fix the features znode or something. I don't know.

Reply:
For this PR I'm trying to minimize functional changes to the ZK controller. If we end up actually using feature flags in ZK mode we can re-evaluate some of these things, but for now I'm trying to reduce impact

Reply:
Yeah, that's fair. We can follow up on this later if needed.